### PR TITLE
feat: add publish & update action

### PR DESCRIPTION
**Pull request by joshuaellis**
### What does it do?

* adds addDocumentAction API
* adds publish & update action
* refactors DescriptionComponentRenderer for better rendering (less popping, more wait till everyone is ready)
* adds published tab

### How to test it?

* You should be able to create & save an entry
* You should be able to publish that entry

### Related issue(s)/PR(s)

* resolves CONTENT-2018
* resolves CONTENT-2019
* resolves CONTENT-2020
* resolves CONTENT-2022
* resolves CONTENT-2023
* resolves CONTENT-2113

### CODE REVIEW

**Comment by joshuaellis** [comment]
The API tests are broken on the feature branch, we'll fix them there :)

File: packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx [hunk 1]
```diff
@@ -17,74 +20,76 @@ interface DescriptionComponent<Props, Description> {
...
 interface DescriptionComponentRendererProps<Props = any, Description = any> {
-  children: (descriptions: Description[]) => React.ReactNode;
+  children: (descriptions: Array<Description & { id: string }>) => React.ReactNode;
   descriptions: DescriptionComponent<Props, Description>[];
   props: Props;
 }
 
-const DescriptionComponentRendererImpl = <Props, Description>({
+const DescriptionComponentRenderer = <Props, Description>({
   children,
```
Comment (remidej) [inline packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx:30]: It would be nice to have a few comments in this component because none of it is obvious to me

File: packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx [hunk 2]
```diff
@@ -17,74 +20,76 @@ interface DescriptionComponent<Props, Description> {
...
+
+  const throttledRequestUpdate = useThrottledCallback(requestUpdate, 60, { trailing: true });
+
+  const update = React.useCallback<DescriptionProps<Props, Description>['update']>(
+    (id, description) => {
+      if (description === null) {
+        delete statesRef.current[id];
+      } else {
+        const current = statesRef.current[id];
+        statesRef.current[id] = { ...current, value: { ...description, id } };
```
Comment (remidej) [inline packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx:58]: Why not mutate `statesRef.current[id].value` directly?

File: packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx [hunk 3]
```diff
@@ -17,74 +20,76 @@ interface DescriptionComponent<Props, Description> {
...
+    [descriptions]
+  );
+
+  const states = React.useMemo(
+    () =>
+      ids
+        .map((id) => statesRef.current[id]?.value)
+        .filter((state) => state !== null && state !== undefined),
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- we leave tick in the deps to ensure the memo is recalculated
+    [ids, tick]
```
Comment (remidej) [inline packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx:77]: Maybe this is necessary but it seems a bit sketchy. I guess I'm wondering why we need to use forceUpdate at all, instead of trusting React to do state=>ui?

File: packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx [hunk 4]
```diff
@@ -134,5 +144,20 @@ function getCompId<T, K>(comp: DescriptionComponent<T, K>): string {
   return id;
 }
 
+const useShallowCompareMemoize = <T,>(value: T): Array<T | undefined> => {
```
Comment (remidej) [inline packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx:147]: The coma after T seems forgotten

File: packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx [hunk 5]
```diff
@@ -134,5 +144,20 @@ function getCompId<T, K>(comp: DescriptionComponent<T, K>): string {
...
+  const ref = React.useRef<T | undefined>(undefined);
+
+  if (!isEqual(value, ref.current)) {
+    ref.current = value;
+  }
+
+  return [ref.current];
+};
+
+const useShallowCompareEffect = (callback: React.EffectCallback, dependencies: any) => {
```
Comment (remidej) [inline packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx:157]: We at least know that dependencies is an array (or undefined)

File: packages/core/admin/admin/src/content-manager/pages/EditView/components/DocumentActions.tsx [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        <Button
+          flex={1}
+          startIcon={primaryAction.icon}
+          disabled={primaryAction.disabled}
+          onClick={primaryAction.onClick}
+          justifyContent="center"
+        >
+          {primaryAction.label}
+        </Button>
+        {restActions.length > 0 ? (
```
Comment (remidej) [inline packages/core/admin/admin/src/content-manager/pages/EditView/components/DocumentActions.tsx:97]: restActions are not actually implemented yet right? I don't see the value being used anywhere

File: packages/core/admin/admin/src/content-manager/pages/EditView/components/DocumentActions.tsx [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    onClick: async () => {
+      if (id) {
+        const res = await publish({
+          collectionType,
+          model,
+          id,
+        });
+
+        if ('data' in res) {
+          navigate({ search: `?status=published` });
```
Comment (remidej) [inline packages/core/admin/admin/src/content-manager/pages/EditView/components/DocumentActions.tsx:161]: Shouldn't we make sure we keep the locale here?

File: packages/core/admin/admin/src/content-manager/pages/EditView/components/DocumentActions.tsx [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+          const res = await create(
+            {
+              model,
+            },
+            document
+          );
+
+          if ('data' in res) {
+            /**
+             * TODO: refactor the router so we can just do `../${res.data.id}` instead of this.
```
Comment (remidej) [inline packages/core/admin/admin/src/content-manager/pages/EditView/components/DocumentActions.tsx:219]: 💯 we have the same issue with the history page's back button

File: packages/core/admin/admin/src/core/apis/content-manager.ts [hunk 9]
```diff
@@ -28,6 +32,10 @@ interface EditViewContext {
    * does not have draft & published enabled.
    */
   activeTab: 'draft' | 'published' | null;
+  /**
+   * Will be either 'single-types' | 'collection-types'
+   */
+  collectionType: string;
```
Comment (remidej) [inline packages/core/admin/admin/src/core/apis/content-manager.ts:38]: Why not use a union then?

**Review by remidej — COMMENTED** [review]
https://i.giphy.com/GbddJJKbqtY2j8QRKq.webp

File: packages/core/admin/admin/src/content-manager/pages/EditView/EditViewPage.tsx [hunk 10]
```diff
@@ -132,89 +152,66 @@ const EditViewPage = () => {
...
-      <TabGroup variant="simple" label="Document version">
+      <Header
+        isCreating={isCreatingDocument}
+        status={getDocumentStatus(document, meta)}
+        title={documentTitle}
+      />
+      <TabGroup
+        ref={tabApi}
+        variant="simple"
+        label="Document version"
```
[the remark this answers is not in the document]
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/pages/EditView/EditViewPage.tsx:176]: Yes, it should be 👍🏼

File: packages/core/admin/admin/src/content-manager/pages/EditView/components/DocumentActions.tsx (hunk 7 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/pages/EditView/components/DocumentActions.tsx:161]: I've not implemented i18n for this branch yet, i'll do all that at that point :)

File: packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx (hunk 4 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx:147]: This is the same lint error you've asked about before, in `tsx` files it assumes its JSX which it's not :(

File: packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx (hunk 3 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx:77]: I'll write a better description, thanks for pointing it out! 

TLDR; the keys won't change if we don't get new descriptions, but the state refs could do, hence why we force update to re-evaluate the ref.

File: packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx (hunk 1 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx:30]: Yes, i had tests but they were very flakey and impossible to be reliable that would have helped. I'll write comments :)

File: packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx (hunk 2 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/components/DescriptionComponentRenderer.tsx:58]: FP over OOP i'd say :)

File: packages/core/admin/admin/src/content-manager/pages/EditView/components/DocumentActions.tsx (hunk 8 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/pages/EditView/components/DocumentActions.tsx:219]: Nice, one of us could do this in the cleanup phase of v5 👍🏼

**Review by markkaylor — APPROVED** [review]
Tested locally, it works. Left some comments but I think they are all more or less optional. I'd at least commit the typos though 🙃

File: packages/core/admin/admin/src/content-manager/pages/EditView/EditViewPage.tsx
[the remark this answers is not in the document]
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/pages/EditView/EditViewPage.tsx:1]: It's really interesting you bring this up, because i'm wondering if we should be moving to match our URL better which might even be clearer, it's then:

```shell
- content-manager
- - :collection-type
- - - :model
- - - - :id
```

It'd be great to universally align on this process and force it across the monorepo, because i can see we're all doing slightly different things. 🤔

File: packages/core/admin/admin/src/content-manager/components/Form.tsx [hunk 11]
```diff
@@ -56,179 +61,299 @@ interface FormProps<TFormValues extends FormValues = FormValues>
...
+                return setIn(errors, err.path!, err.message);
+              }
+              for (const error of err.inner) {
+                if (!getIn(errors, error.path!)) {
+                  errors = setIn(errors, error.path!, err.message);
+                }
+              }
+            }
+
+            return errors;
```
[the remark this answers is not in the document]
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/components/Form.tsx:115]: I'll keep it in mind & look out for reconciliation, for now i won't be over-eager and move it out 🤔

File: packages/core/admin/admin/src/content-manager/pages/EditView/EditViewPage.tsx [hunk 12]
```diff
@@ -132,89 +152,69 @@ const EditViewPage = () => {
...
+        title={documentTitle}
+      />
+      <TabGroup
+        ref={tabApi}
+        variant="simple"
+        label={formatMessage({
+          id: getTranslation('containers.edit.tabs.label'),
+          defaultMessage: 'Document status',
+        })}
+        initialSelectedTabIndex={!status || status === 'draft' ? 0 : 1}
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/pages/EditView/EditViewPage.tsx:180]:
```suggestion
        initialSelectedTabIndex={status === 'published' ? 1 : 0}
```
